### PR TITLE
code refactoring for tenant management API

### PR DESCRIPTION
**Pull request by lipingxue**
This PR includes:
1. Refactor code to add an abstract layer provide APIs for tenant management
2. add vmdk_ops unit test for multi-tenency
3. make "name" unique in "tenants" table

### CODE REVIEW

**Comment by lipingxue** [comment]
Please review it @msterin @shaominchen

**Comment by pdhamdhere** [comment]
@lipingxue CI failing. Please fix.

File: esx_service/cli/vmdkops_admin.py [hunk 1]
```diff
@@ -811,66 +764,47 @@ def generate_tenant_ls_rows(tenant_list):
...
-    return None, vms
-
 def tenant_create(args):
     """ Handle tenant create command """
-    error_info, vms = generate_tuple_from_vm_list(args.vm_list)
-    if error_info:
-        return operation_fail(error_info)
-      
-    name = args.name 
+    name = args.name
```
Comment (pdhamdhere) [inline esx_service/cli/vmdkops_admin.py]: Currently we do not validate any parameters. We should. Separate PR is fine with me.

File: esx_service/utils/auth.py [hunk 2]
```diff
@@ -324,6 +326,25 @@ def add_volume_to_volumes_table(tenant_uuid, datastore, vol_name, vol_size_in_MB
 
     return None
 
+def remove_volume_from_volumes_table(tenant_uuid, datastore, vol_name):
+    """ Remove volume from volumes table. """
```
Comment (pdhamdhere) [inline esx_service/utils/auth.py]: Can we add a line about return code e.g. "Returns None on success or error string" to all functions?

File: esx_service/utils/auth_api.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+def generate_tuple_from_vm_list(vm_list):
+    """ Generate a list of (vm_uuid, vm_name) pair """
+    if not vm_list:
+        return None, []
+    vms = []
+    for vm_name in vm_list:
+        vm_uuid = vmdk_utils.get_vm_uuid_by_name(vm_name)
+        if not vm_uuid:
+            error_info =  "Cannot find vm_uuid for vm {0}, tenant_create failed".format(vm_name)
+            return error_info, None
```
Comment (pdhamdhere) [inline esx_service/utils/auth_api.py]: I think I had commented on this last PR. We should not terminate loop on first bad VM name. Rather iterate through entire list and return VMs that are successfully assigned and VMs that couldn't be found.

File: esx_service/utils/auth_api.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+        vms.append((vm_uuid, vm_name))
+    
+    return None, vms
+
+def default_privileges():
+     privileges = [{'datastore': 'datastore1',
+                     'global_visibility': 0,
+                      'create_volume': 0,
+                      'delete_volume': 0,
+                     'mount_volume': 0,
```
Comment (pdhamdhere) [inline esx_service/utils/auth_api.py]: nit: Looks off by a space.

File: esx_service/utils/auth_api.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+            error_info =  "Cannot find vm_uuid for vm {0}, tenant_create failed".format(vm_name)
+            return error_info, None
+        vms.append((vm_uuid, vm_name))
+    
+    return None, vms
+
+def default_privileges():
+     privileges = [{'datastore': 'datastore1',
+                     'global_visibility': 0,
+                      'create_volume': 0,
```
Comment (pdhamdhere) [inline esx_service/utils/auth_api.py]: Are we going to group privileges as discussed in #709

File: esx_service/utils/auth.py [hunk 6]
```diff
@@ -28,6 +28,7 @@
 CMD_REMOVE = 'remove'
 CMD_ATTACH = 'attach'
 CMD_DETACH = 'detach'
+CMD_LIST = 'list'
```
Comment (pdhamdhere) [inline esx_service/utils/auth.py]: I had removed this in my last PR. Any reason we are re-introducing? 
See executeRequest. Auth checks are bypassed for 'list' command.

File: esx_service/utils/auth_api.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    return privileges
+
+def generate_privileges_dict(privileges):
+    # privileges is a list which is read from auth DB
+    # it has the following format
+    # (tenant_uuid, datastore, global_visibility, create_volume, delete_volume,
+    # mount_volume, max_volume_size, usage_quota)
+    privileges_dict = {}
+    privileges_dict[auth_data_const.COL_DATASTORE] = privileges[1]
+    privileges_dict[auth_data_const.COL_GLOBAL_VISIBILITY] = privileges[2]
```
Comment (pdhamdhere) [inline esx_service/utils/auth_api.py]: Didn't we get rid of GLOBAL_VISIBILITY?

File: esx_service/utils/auth_data.py [hunk 8]
```diff
@@ -359,15 +359,16 @@ def create_tables(self):
                 '''
                     CREATE TABLE tenants(
                     -- uuid for the tenant, which is generated by create_tenant() API
-                    id TEXT PRIMARY KEY NOT NULL,
+                    id TEXT NOT NULL,
```
Comment (pdhamdhere) [inline esx_service/utils/auth_data.py]: If name is unique then why do we need ID?

File: esx_service/vmdk_ops.py [hunk 9]
```diff
@@ -177,12 +177,27 @@ def createVMDK(vmdk_path, vm_name, vol_name, opts={}):
...
+   
     backing, needs_cleanup = get_backing_device(vmdk_path)
     cleanup_backing_device(backing, needs_cleanup)
 
+    # create succeed, insert the volume information into "volumes" table
+    if tenant_uuid:
+        vol_size_in_MB = convert.convert_to_MB(auth.get_vol_size(opts))
+        auth.add_volume_to_volumes_table(tenant_uuid, datastore, vol_name, vol_size_in_MB)
+    else:
+        logging.warning(" VM %s does not belong to any tenant", vm_name)
```
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: warnings can flood logs. How about changing it to debug?

File: esx_service/vmdk_ops.py [hunk 10]
```diff
@@ -427,9 +442,15 @@ def removeVMDK(vmdk_path):
         elif rc != 0:
             return err("Failed to remove %s. %s" % (vmdk_path, out))
         else:
+            # remove succeed, remove infomation of this volume from volumes table
+            if tenant_uuid:
+                error_info = auth.remove_volume_from_volumes_table(tenant_uuid, datastore, vol_name)
+                return error_info
+            else:
+                if not vm_name:
+                    logging.warning("VM %s does not belong to any tenant", vm_name)
```
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: Same here. These warnings can flood logs. Change it to debug?

File: esx_service/vmdk_ops_test.py [hunk 11]
```diff
@@ -532,10 +549,296 @@ def test_vmdkop_authorize(self):
...
+    
+
+    def setUp(self):
+        """ Setup run before each test """
+        logging.debug("VMDKTenantTest setUp path =%s", path)
+        
+        if (not self.datastore_name):
+            datastores = vmdk_utils.get_datastores()
+            datastore = datastores[0]
+            if (not datastore):
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: this looks suspicious. Is there a case where non-empty array with None as the first element is returned by get_datastores() ?

File: esx_service/vmdk_ops_test.py [hunk 12]
```diff
@@ -532,10 +549,296 @@ def test_vmdkop_authorize(self):
...
+                                                    description = description, 
+                                                    default_datastore = default_datastore, 
+                                                    default_privileges = default_privileges, 
+                                                    vm_list = vm_list, 
+                                                    privileges = privileges)
+        self.assertEqual(None, error_info)      
+          
+    def tearDown(self):
+        """ Cleanup after each test """
+        logging.debug("VMDKTenantTest  tearDown path")
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: logging.info ? we are in tests, it's important to have traces

File: esx_service/vmdk_ops_test.py [hunk 13]
```diff
@@ -532,10 +549,296 @@ def test_vmdkop_authorize(self):
...
+            
+    def cleanup(self):
+        # cleanup existing non-tenant volume
+        if not self.datastore_path:
+            vmdk_path = vmdk_utils.get_vmdk_path(self.datastore_path, self.non_tenant_vol_name)
+            if vmdk_ops.getVMDK(vmdk_path, self.non_tenant_vol_name, self.datastore):
+                vmdk_ops.removeVMDK(vmdk_path)
+
+        # cleanup existing tenant
+        error_info = auth_api._tenant_rm(
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: nit: PEP8 (just a reminder). no spaces around `=` in param assignment

File: esx_service/vmdk_ops_test.py [hunk 14]
```diff
@@ -532,10 +549,296 @@ def test_vmdkop_authorize(self):
...
+        # list volumes for tenant2
+        opts = {}
+        result = vmdk_ops.executeRequest(vm2_uuid, self.vm2_name, self.vm2_config_path, auth.CMD_LIST, None, opts)
+       
+        self.assertEqual(3, len(result))
+        self.assertEqual("tenant2_vol1", result[0]['Name'])
+        self.assertEqual("tenant2_vol2", result[1]['Name'])
+        self.assertEqual("tenant2_vol3", result[2]['Name'])
+
+
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: A general comment on tests: it is better to have them data drivel . I.e. instead of using tenantx_volx ad opts and other stuff in the code, ALL that should be in data structures and the test should just taverse the structures. See example (testInfo traversing) in vmdk_ops_test.py.

It's OK to leave it as is for this PR but please do open an issue for test rearrangement and making them data driven - that allows to understand tests better and to add to the test set easily

File: esx_service/vsan_policy_test.py [hunk 15]
```diff
@@ -62,6 +62,7 @@ def test_create_delete(self):
 
     def test_delete_nonexistent_policy_fails(self):
         self.assertNotEqual(None, vsan_policy.delete(self.name))
+        print ("The test itself is expected to fail, and please ignore the errors printed above.")
```
Comment (msterin) [inline esx_service/vsan_policy_test.py]: does not it have to be stderr ? Or log ? (just asking  since I do not see this message in the log

File: esx_service/utils/auth.py (hunk 6 above)
Comment (lipingxue) [inline esx_service/utils/auth.py]: In my unit test code, I need to test 'list' command. It seems that your change assume we only put commands which need authorize in auth.py. I will remove 'list' command from this file and update the comment.

File: esx_service/utils/auth_api.py (hunk 5 above)
Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Existing issue #709 tracked this.

File: esx_service/cli/vmdkops_admin.py (hunk 1 above)
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.py]: Issue https://github.com/vmware/docker-volume-vsphere/issues/734 is filed to track this.

File: esx_service/utils/auth_api.py (hunk 3 above)
Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Done.

File: esx_service/utils/auth_api.py (hunk 4 above)
Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Done.

File: esx_service/utils/auth_data.py (hunk 8 above)
Comment (lipingxue) [inline esx_service/utils/auth_data.py]: PR #https://github.com/vmware/docker-volume-vsphere/issues/711 tracks "tenant rename".

File: esx_service/vmdk_ops.py (hunk 9 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Change the log to "DEBUG".

File: esx_service/vmdk_ops.py (hunk 10 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: In this PR, I just define a global for format string. But in the long run, I think we need to have a file which defines all the error code and corresponding error string for all the error message we used in our code.

File: esx_service/utils/auth_api.py (hunk 7 above)
Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Done. Remove "global_visibility" from the code.

File: esx_service/cli/vmdkops_admin.py [hunk 16]
```diff
@@ -811,66 +764,47 @@ def generate_tenant_ls_rows(tenant_list):
...
-    error_info, tenant = create_tenant_in_db(
-                                             name = name, 
-                                             description = description, 
-                                             default_datastore = default_datastore, 
-                                             default_privileges = default_privileges, 
-                                             vms = vms, 
-                                             privileges = privileges)
+  
+    error_info, tenant = auth_api._tenant_create(
+                                                 name = name,
```
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.py]: Done.

File: esx_service/cli/vmdkops_admin.py [hunk 17]
```diff
@@ -937,137 +850,46 @@ def generate_tenant_vm_ls_rows(vms):
...
-        return operation_fail(error_info)
-
-    privileges = generate_privileges(args)
-    print privileges
-    error_info = tenant.set_datastore_access_privileges(_auth_mgr.conn, [privileges])
+    name = args.name
+    datastore = args.datastore
+    rights = args.rights
+    volume_maxsize = args.volume_maxsize
+    volume_totalsize = args.volume_totalsize
```
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.py]: Done.

File: esx_service/utils/auth_api.py [hunk 18]
```diff
@@ -0,0 +1,308 @@
+import auth
+import auth_data_const
+import convert
+import auth_data
+import vmdk_utils
+
+def get_auth_mgr():
```
Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Done

File: esx_service/utils/auth_api.py [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+    try:
+        auth_mgr = auth.get_auth_mgr()
+    except auth_data.DbConnectionError, e:
```
Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Done

File: esx_service/utils/auth_api.py [hunk 20]
```diff
@@ -0,0 +1,308 @@
...
+        if not vm_uuid:
+            error_info =  "Cannot find vm_uuid for vm {0}, tenant_create failed".format(vm_name)
+            return error_info, None
+        vms.append((vm_uuid, vm_name))
+    
+    return None, vms
+
+def default_privileges():
+     privileges = [{'datastore': 'datastore1',
+                     'global_visibility': 0,
```
Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Fixed.

File: esx_service/utils/auth_data.py [hunk 21]
```diff
@@ -655,7 +662,11 @@ def _remove_volumes_for_tenant(self, tenant_id):
                 dir_paths.add(vmdk['path'])
                 logging.debug("path=%s filename=%s", vmdk['path'], vmdk['filename'])    
                 logging.debug("Deleting volume path%s", vmdk_path)
-                err = vmdk_ops.removeVMDK(vmdk_path)
+                err = vmdk_ops.removeVMDK(vmdk_path = vmdk_path,
+                                          vol_name = vmdk_utils.strip_vmdk_extension(vmdk['filename']),
+                                          vm_name = None,
+                                          tenant_uuid = tenant_id,
+                                          datastore = vmdk['datastore'])
```
Comment (lipingxue) [inline esx_service/utils/auth_data.py]: Done

File: esx_service/utils/auth_data.py [hunk 22]
```diff
@@ -446,6 +447,12 @@ def create_tenant(self, name, description, default_datastore, default_privileges
 
         """
         logging.debug ("create_tenant name=%s", name)
+        error_info, exist_tenant = self.get_tenant(name)
+        if exist_tenant:
+            error_info = "Tenant {0} already exists".format(name)
+            logging.warning("Tenant %s already exists", name)
```
Comment (lipingxue) [inline esx_service/utils/auth_data.py]: Done.

File: esx_service/utils/auth_api.py [hunk 23]
```diff
@@ -0,0 +1,308 @@
...
+    return error_info
+
+def _tenant_vm_rm(name, vm_list):
+    """ API to remove vms for a tenant """
+    error_info, tenant = get_tenant_from_db(name)
+    if error_info:
+        return error_info
+    
+    if not tenant:
+        error_info = "Tenant {0} does not exist".format(name)
```
Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Done.

File: esx_service/utils/vmdk_utils.py [hunk 24]
```diff
@@ -214,4 +214,29 @@ def get_vm_uuid_by_name(vm_name):
...
+            return path
+    return None
+
+def get_vm_config_path(vm_name):
+    """Returns vm_uuid for given vm_name, or None"""
+    si = vmdk_ops.get_si()
+    try:
+        vm = [d for d in si.content.rootFolder.childEntity[0].vmFolder.childEntity if d.config.name == vm_name]
+        config_path = vm[0].summary.config.vmPathName
+
```
Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: Done.

File: esx_service/vmdk_ops_test.py [hunk 25]
```diff
@@ -287,6 +289,62 @@ def test_failure(self):
             with self.assertRaises(vmdk_ops.ValidationError):
                 vmdk_ops.validate_opts(opts, self.path)
 
+def create_vm(si, vm_name, datastore):
+        content = si.RetrieveContent()
+        datacenter = content.rootFolder.childEntity[0]
+        vm_folder = datacenter.vmFolder
+        hosts = datacenter.hostFolder.childEntity
+        resource_pool = hosts[0].resourcePool
```
Comment (lipingxue) [inline esx_service/vmdk_ops_test.py]: Done

File: esx_service/vmdk_ops_test.py [hunk 26]
```diff
@@ -287,6 +289,62 @@ def test_failure(self):
...
+                                numCPUs=1,
+                                files=vmx_file, 
+                                guestId='rhel5_64Guest', 
+                                version='vmx-11'
+                              )
+
+        task = vm_folder.CreateVM_Task(config=config, pool=resource_pool)
+        vmdk_ops.wait_for_tasks(si, [task])
+
+        vm = [d for d in si.content.rootFolder.childEntity[0].vmFolder.childEntity
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: Badically, SearchIndex.FindChild does exactly the same as your code (scans the folder to match the name) but I suspect it does it on the server, so with many VMs I think it would provide a significant difference in perf. 

```python
name = 'photon1'
dc = si.content.rootFolder.childEntity[0]
si.content.searchIndex.FindChild(dc.vmFolder, name)
```

File: esx_service/vmdk_ops_test.py [hunk 27]
```diff
@@ -287,6 +289,62 @@ def test_failure(self):
...
+                task = vm[0].PowerOnVM_Task()
+                vmdk_ops.wait_for_tasks(si, [task])
+        else:
+            logging.error("Cannot find the test-vm")
+            self.assertFalse(True) 
+    
+def remove_vm(si, vm_name):
+
+    vm = [d for d in si.content.rootFolder.childEntity[0].vmFolder.childEntity 
+            if d.config.name == vm_name]
```
Comment (lipingxue) [inline esx_service/vmdk_ops_test.py]: Done.

File: esx_service/vmdk_ops_test.py (hunk 11 above)
Comment (lipingxue) [inline esx_service/vmdk_ops_test.py]: Done.

File: esx_service/vmdk_ops_test.py (hunk 12 above)
Comment (lipingxue) [inline esx_service/vmdk_ops_test.py]: Done.

File: esx_service/vmdk_ops_test.py (hunk 13 above)
Comment (lipingxue) [inline esx_service/vmdk_ops_test.py]: Done.

File: esx_service/vsan_policy_test.py (hunk 15 above)
Comment (lipingxue) [inline esx_service/vsan_policy_test.py]: put it in logging.info.

File: esx_service/vmdk_ops_test.py [hunk 28]
```diff
@@ -287,6 +289,62 @@ def test_failure(self):
...
+                                   suspendDirectory=None,
+                                   vmPathName=datastore_path)
+
+
+        config = vim.vm.ConfigSpec( 
+                                name=vm_name, 
+                                memoryMB=128, 
+                                numCPUs=1,
+                                files=vmx_file, 
+                                guestId='rhel5_64Guest',
```
Comment (lipingxue) [inline esx_service/vmdk_ops_test.py]: Have to change it back to the original guestId, the create_vm will fail with invalid guestId.

File: esx_service/utils/auth_api.py [hunk 29]
```diff
@@ -0,0 +1,336 @@
+import auth
```
Comment (msterin) [inline esx_service/utils/auth_api.py]: Please add standard file header

File: esx_service/utils/auth_data.py [hunk 30]
```diff
@@ -359,15 +356,16 @@ def create_tables(self):
...
                     -- this field can be changed later by using set_name() API
-                    name TEXT,
+                    name TEXT NOT NULL,
                     -- brief description of the tenant, which is specified by user when creating the tenant
                     -- this field can be changed laster by using set_description API
                     description TEXT,
                     -- not use currently
-                    default_datastore TEXT
+                    default_datastore TEXT,
+                    PRIMARY KEY(id, name)
```
Comment (msterin) [inline esx_service/utils/auth_data.py]: why do we need a key with both id, name  ? does it mean the key changes when name changes ? 
I thought ID would be the key and name would be just a text field (unique is good, but just text, not key) ?

File: esx_service/utils/auth_data.py [hunk 31]
```diff
@@ -446,6 +442,12 @@ def create_tenant(self, name, description, default_datastore, default_privileges
 
         """
         logging.debug ("create_tenant name=%s", name)
+        error_info, exist_tenant = self.get_tenant(name)
+        if exist_tenant:
+            error_info = error_code.TENANT_ALREADY_EXIST.format(name)
```
Comment (msterin) [inline esx_service/utils/auth_data.py]: nice !

File: esx_service/utils/error_code.py [hunk 32]
```diff
@@ -0,0 +1,7 @@
+# Tenant related error message
```
Comment (msterin) [inline esx_service/utils/error_code.py]: Please add standard file header

File: esx_service/utils/vmdk_utils.py [hunk 33]
```diff
@@ -214,4 +214,38 @@ def get_vm_uuid_by_name(vm_name):
         return vm[0].config.uuid
     except:
         return None
+def get_datastore_path(ds_name):
+    for (datastore, url_name, path) in get_datastores():
```
Comment (msterin) [inline esx_service/utils/vmdk_utils.py]: dosctring (or comment) is missing. Also, an empty line before `def`  is needed

File: esx_service/utils/auth_api.py (hunk 29 above)
Comment (pdhamdhere) [inline esx_service/utils/auth_api.py]: I thought it was auto-generated? CC/ @kerneltime

File: esx_service/utils/auth_api.py [hunk 34]
```diff
@@ -0,0 +1,336 @@
...
+    error_info, tenant = create_tenant_in_db(
+                                             name=name, 
+                                             description=description, 
+                                             default_datastore=default_datastore, 
+                                             default_privileges=default_privileges, 
+                                             vms=vms, 
+                                             privileges=privileges)
+    if error_info:
+        return error_info, None
+
```
Comment (pdhamdhere) [inline esx_service/utils/auth_api.py]: Print/ Log `not_found_vms` informing user?

File: esx_service/utils/auth_data.py (hunk 21 above)
Comment (pdhamdhere) [inline esx_service/utils/auth_data.py]: Spaces still exist. Forgot to push?

File: esx_service/utils/error_code.py [hunk 35]
```diff
@@ -0,0 +1,7 @@
...
+VM_NOT_BELONG_TO_TENANT = "VM {0} does not belong to any tenant"
+TENANT_NOT_EXIST = "Tenant {0} does not exist"
+TENANT_ALREADY_EXIST = "Tenant {0} already exists"
+
```
Comment (pdhamdhere) [inline esx_service/utils/error_code.py]: Nice!

File: esx_service/vmdk_ops.py [hunk 36]
```diff
@@ -177,12 +178,25 @@ def createVMDK(vmdk_path, vm_name, vol_name, opts={}):
     if not create_kv_store(vm_name, vmdk_path, opts):
         msg = "Failed to create metadata kv store for {0}".format(vmdk_path)
         logging.warning(msg)
-        removeVMDK(vmdk_path)
-        return err(msg)
-
+        error_info = err(msg)
+        remove_err = removeVMDK(vmdk_path = vmdk_path,
```
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: remove space?

File: esx_service/vmdk_ops.py [hunk 37]
```diff
@@ -626,18 +648,20 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
...
-        # create succeed, insert infomation of this volume to volumes table
-        if not response:
-            if tenant_uuid:
-                vol_size_in_MB = convert.convert_to_MB(auth.get_vol_size(opts))
-                auth.add_volume_to_volumes_table(tenant_uuid, datastore, vol_name, vol_size_in_MB)
-            else:
-                logging.warning(" VM %s does not belong to any tenant", vm_name)
-
+    elif cmd == "create":          
+        response = createVMDK(vmdk_path = vmdk_path,
```
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: remove spaces?

File: esx_service/vmdk_ops.py [hunk 38]
```diff
@@ -626,18 +648,20 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
...
+        response = createVMDK(vmdk_path = vmdk_path, 
+                              vm_name = vm_name, 
+                              vol_name = vol_name, 
+                              opts = opts, 
+                              tenant_uuid = tenant_uuid, 
+                              datastore = datastore)
+       
     elif cmd == "remove":
-        response = removeVMDK(vmdk_path)
+        response = removeVMDK(vmdk_path = vmdk_path,
```
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: remove spaces?

File: esx_service/utils/auth_api.py (hunk 29 above)
Comment (msterin) [inline esx_service/utils/auth_api.py]: It was auto-generated when we did a pass of adding headers :-). Now it is copy-n-paste before push.

Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Done.

File: esx_service/utils/auth_data.py (hunk 30 above)
Comment (lipingxue) [inline esx_service/utils/auth_data.py]: Done. Make "name" to be UNIQUE, but not the key.

File: esx_service/utils/error_code.py (hunk 32 above)
Comment (lipingxue) [inline esx_service/utils/error_code.py]: Done.

File: esx_service/utils/auth_api.py (hunk 34 above)
Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Done.

File: esx_service/utils/error_code.py (hunk 35 above)
Comment (lipingxue) [inline esx_service/utils/error_code.py]: Done.

File: esx_service/vmdk_ops.py (hunk 36 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Done.

File: esx_service/vmdk_ops.py (hunk 37 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Done.

File: esx_service/vmdk_ops.py (hunk 38 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Done.

**Comment by msterin** [comment]
@lipingxue - it looks like there are some conflicts to be resolved. and the latest changes (marked as "done" are still not in. Looking forward...

**Comment by lipingxue** [comment]
Yes.  After rebase with the master, there are some conflict with latest change in master(Bruno's change). I am addressing those conflict and rerun the test. Will do the push when it is done.

**Comment by lipingxue** [comment]
@msterin I have resolved conflict  and push the change. Please take another look.

File: esx_service/vmdk_ops_test.py [hunk 39]
```diff
@@ -438,52 +438,6 @@ def remove_vm(si, vm_name):
         task = vm.Destroy_Task()
         vmdk_ops.wait_for_tasks(si, [task])
 
-class ValidationTestCase(unittest.TestCase):
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: why this test is removed?

**Comment by lipingxue** [comment]
@msterin 
As we discussed last Friday, CI test failed due to DB schema existed in the testbed.
In order to make the test always start from a clean state, we decided to add code to remove auth_DB at each test. I added code for that, but it does not work.
The CI test failed with  "disk I/O error"  when accessing auth DB

I checked the log and also looked at our code, and I figured out this is reason why I see the above error.  We call function "get_auth_mgr()" to get a connection to the DB.  This function will call "thread_local._auth_mgr.connect()" ONLY if it does not have a cached value. In function"AuthorizationDataManager.connect()", we have the logic to decide whether we need to create table or not (check whether DB file exists or not).

Assume we start from the state that we have pre-exist DB in the testbed.
1.  first time call "get_auth_mgr" to get a DB connection. Since this is the first time, it will further call AuthorizationDataManager.connect() to check whether create table is needed or not. Here
we don't need create the table since we already has the DB file in the test bed. The function 
return a connection for the DB.

2. remove the auth DB file in the test code

3. second time call "get_auth_mgr" to get a DB connection. This time, it will not call AuthorizationDataManager.connect() since it already has the cached value. Therefore, it returned
a connection to staled DB file.


I think we only need to remove the old DB one time. This logic also apply to the possible DB migration.
Now I change the logic as follows:
1. At the beginning of  AuthorizationDataManager.connect(), check if we have DB file, it we have, and we haven't removed it (upgraded), do remove the DB.
2. After that, existing code will check whether we have the DB file (we don't have now since we just remove it),  and if we don't have, it will create the DB file and return the connection to the DB.

Please review this  newly added code.

File: esx_service/utils/auth_data.py [hunk 40]
```diff
@@ -308,6 +310,22 @@ def get_auth_db_path(self):
...
+        # whether migrate is needed
+        return True
+    
+    def migrate_auth_db(self):
+        # currently, just remove the existing DB file
+        logging.debug("migrate_auth_db")
+        logging.debug("auth DB %s exist, try to remove the file", AUTH_DB_PATH)
+        os.remove(AUTH_DB_PATH)
+        # insert some delay 
+        time.sleep(60)
```
Comment (msterin) [inline esx_service/utils/auth_data.py]: why the delay ?

File: esx_service/utils/auth_data.py [hunk 41]
```diff
@@ -319,6 +337,14 @@ def connect(self):
         Raises a ConnectionFailed exception when connect fails.
 
         """
+        global has_migrate_auth_db
+
+        if os.path.isfile(self.db_path):
+            # already exist
+            if not has_migrate_auth_db and self.need_migrate_auth_db():
```
Comment (msterin) [inline esx_service/utils/auth_data.py]: self.need_migrate_auth_db() need to check for 'has already migrated' (what you cann  has_migrate_auth_db ) for better encapsulation

File: esx_service/utils/auth_data.py [hunk 42]
```diff
@@ -319,6 +337,14 @@ def connect(self):
...
+                self.migrate_auth_db()
+                has_migrate_auth_db = True
```
Comment (msterin) [inline esx_service/utils/auth_data.py]: well, since it's global it's "test and set" should be atomic , so I suggest a lock here

File: misc/scripts/deploy-tools.sh [hunk 43]
```diff
@@ -117,9 +117,11 @@ function deployesx {
     do
         TARGET=root@$ip
         log "Deploying to ESX $TARGET"
+        $SSH $TARGET  rm -f /etc/vmware/vmdkops/log_config.json
```
Comment (msterin) [inline misc/scripts/deploy-tools.sh]: This one should be a different commit - it's a useful fix

**Comment by lipingxue** [comment]
@msterin I have addressed your comments, please review.

**Comment by lipingxue** [comment]
Fixed #710
